### DISCUSSION TITLE
How to "shamefully hoist" a package with two dependent packages depending on a third package with two different versions?

### DISCUSSION THREAD

**Discussion by obones** [pnpm/pnpm#7902] [Q&A]
Hello,

I have a monorepo where one of the packages has the following `package.json`:

```
{
  "name": "archiver",
  "dependencies": {
    "archiver": "5.3.0"
  },
  "scripts": {
    "prepack": "pnpm --config.shamefully-hoist=true --config.shared-workspace-lockfile=false deploy --filter archiver --prod nodejs --force",
    "pack": "npx --yes npm-pack-zip --destination ../",
    "postpack": "npx --yes rimraf nodejs"
  }
}
```

When running `pnpm run pack`, I get a nice `archiver.zip` in the parent folder which I can later use as the source for an AWS lambda layer.
I have done this with other packages and it worked fine for all of them, except for the one above where it fails with this message:

```
Error: Cannot find module 'readable-stream/passthrough'
Require stack:
- /opt/nodejs/node_modules/lazystream/lib/lazystream.js
- /opt/nodejs/node_modules/archiver-utils/index.js
- /opt/nodejs/node_modules/archiver/lib/core.js
- /opt/nodejs/node_modules/archiver/index.js
- /var/task/index.js
```

This was very surprising to me as the `readable-stream` library is placed inside the hoisted structure just fine.
Before using pnpm, I had done a simple export with `npm` and created the zip file manually. This allowed me to spot one difference:

- Inside the `lazystream` package folder, there was a ` node_modules/readable-stream` folder which was now missing from my newly generated zip.

Looking at the dependencies, I see two things:

1. `archiver` 5.3.0 [depends](https://github.com/archiverjs/node-archiver/blob/692be046cf17fd185676984fa51c27d4957b6ea3/package.json) on `readable-stream` version 3.6 
2. `archiver-utils` depends on `lazystream` which hasn't changed in ages and still [depends](https://github.com/jpommerening/node-lazystream/blob/master/package.json#L35) on `readable-stream` version 2.0.5

This explains why there was a `node_modules` folder specifically for `lazystream` but I could not find an option to be given to `pnpm` above to replicate this structure when two packages depend on a different version of the same package.

I was able to come up with a workaround by using lots of different commands and ended up with this:

```
{
  "name": "archiver",
  "dependencies": {
    "archiver": "5.3.0"
  },
  "scripts": {
    "prepack:1": "pnpm --config.shamefully-hoist=true --config.shared-workspace-lockfile=false deploy --filter archiver --prod nodejs_links --force",
    "prepack:2": "npx --yes cpy-cli \"./nodejs_links/node_modules/.pnpm/readable-stream@2.3.8/node_modules/readable-stream/**\" ./nodejs_links/node_modules/lazystream/node_modules/readable-stream/",
    "prepack:3": "npx --yes cpy-cli \"./nodejs_links/**\" ./nodejs",
    "prepack:4": "npx --yes rimraf nodejs_links nodejs/node_modules/.pnpm",
    "prepack": "pnpm run prepack:1 && pnpm run prepack:2 && pnpm run prepack:3 && pnpm run prepack:4",
    "pack": "npx --yes cross-replace npx --yes bestzip ../$npm_package_name.zip nodejs/*",
    "postpack": "npx --yes rimraf nodejs"
  }
}
```

But while this fixes my issue, it does not feel very extensible should another dependency need to be "fixed" in a similar way.

As I was not sure this is a bug in `pnpm` itself, I preferred to start a discussion here, but please do not hesitate to let me know if I should move this to the issue tracker.

**Comment by zkochan** [comment]
Try to use `--config.node-linker=hoisted` instead of `--config.shamefully-hoist=true`

**Reply by obones** [reply]
Thanks, but I must have missed something because the created `nodejs` folder only contains the following structure:
```
nodejs
   |- package.json
   \ node_modules
      |- .modules.yaml
      \ .pnpm
         - lock.yam
```

And so the final zip does not contain any of the modules sources.
